### DISCUSSION TITLE
Headless rendering in node with custom component types

### DISCUSSION THREAD

**Discussion by derikb** [GrapesJS/grapesjs#6223] [Q&A]
Trying to set it up so node can render the html etc from the grapesjs JSON we store from the editor (based on the comments in https://github.com/GrapesJS/grapesjs/discussions/3672 ). I can get things to work with the headless config option _if_ I don't try to use any custom component types. (I get errors for the types not being found, but I get basic html output.)

When I add a call to `editor.DomComponents.addType` I start getting errors related to  document being undefined in BrowserParserCss method. I don't see a way to get the grapes import to use an alternate document (like from JSDOM).

Is it possible to use custom Component Types server side, or am I just wasting my time.

**Comment by artf** [comment]
You should replace the built-in browser parser https://github.com/GrapesJS/grapesjs/issues/5232
